WalletWasabi/Extensions/StreamExtensions.cs

```diff
@@ -1,29 +1,54 @@
+using System.Buffers;
 using System.Threading;
 using System.Threading.Tasks;
-using System;
 
 namespace System.IO
 {
 	public static class StreamExtensions
 	{
 		public static async Task<int> ReadByteAsync(this Stream stream, CancellationToken ctsToken = default)
 		{
-			var buf = new byte[1];
-			var len = await stream.ReadAsync(buf.AsMemory(0, 1), ctsToken).ConfigureAwait(false);
-			if (len == 0)
+			ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+			byte[] buffer = pool.Rent(1);
+			try
 			{
-				return -1;
-			}
+				int len = await stream.ReadAsync(buffer.AsMemory(0, 1), ctsToken).ConfigureAwait(false);
+
+				// End of stream.
+				if (len == 0)
+				{
+					return -1;
+				}
 
-			return buf[0];
+				return buffer[0];
+			}
+			finally
+			{
+				pool.Return(buffer);
+			}
 		}
 
-		public static async Task<int> ReadBlockAsync(this Stream stream, byte[] buffer, int count, CancellationToken ctsToken = default)
+		/// <summary>
+		/// Attempts to read <paramref name="count"/> bytes from <paramref name="stream"/>.
+		/// </summary>
+		/// <param name="stream">Stream to read from.</param>
+		/// <param name="buffer">Buffer whose length must be at least <paramref name="count"/> elements.</param>
+		/// <param name="count">Number of bytes to read.</param>
+		/// <param name="cancellationToken">Cancellation token to cancel the asynchronous operation.</param>
+		/// <returns>Number of read bytes. At most <paramref name="count"/>.</returns>
+		public static async Task<int> ReadBlockAsync(this Stream stream, byte[] buffer, int count, CancellationToken cancellationToken = default)
 		{
-			var left = count;
+			int left = count;
 			while (left != 0)
 			{
-				var read = await stream.ReadAsync(buffer.AsMemory(count - left, left), ctsToken).ConfigureAwait(false);
+				int read = await stream.ReadAsync(buffer.AsMemory(count - left, left), cancellationToken).ConfigureAwait(false);
+
+				// End of stream.
+				if (read == 0)
+				{
+					break;
```

Comment:
You mentioned that you are planning to fix one edge case in the next PR. Are you planning to throw an exception here if the packet is not finished but we are not receiving any more data?

Reply:
> Are you planning to throw an exception here if the packet is not finished but we are not receiving any more data?

We currently call `ReadBlockAsync` then we check return value to throw or not throw an exception. But those exception types are different. So if we throw an exception at the `break` place, we would re-throw or wrap it on an upper level immediately, that seems wasteful. Secondly, these networking functions follow a certain pattern of "I have read this many bytes" and it looks to me that our `ReadBlockAsync` is consistent with that .NET pattern and it's documented too so it should not be confusing.

Reply:
ACK!